common/models/transducer/multi_stager.py

```diff
@@ -0,0 +1,284 @@
+                                                "same_dim_tags_as": {"t": _output_len_tag}}
+  return net
+
+
+def get_fixed_path_meta_dataset(task: str,
+                                path_2_hdf: str,
+                                ctx: Context):
+  """
+  TODO:
+  """
+  train = ctx.train
+
+  # TODO should be put in a metadataset together with the normal dataset
+  align_dataset = {
+    "class": "HDFDataset", "files": [path_2_hdf],
+    "use_cache_manager": True,
+    # "unique_seq_tags": True  # dev set can exist multiple times
+    # TODO: otherwise not right selection
+    # "seq_list_filter_file": files["segments"],
+    # "partition_epoch": epoch_split,
+    # TODO: do we really need the num_seq
+    # "estimated_num_seqs": (estimated_num_seqs[data] // epoch_split) if data in estimated_num_seqs else None,
+  }
+  if train:
+    # TODO: do we really need the num_seq
+    # align_dataset["seq_ordering"] = "laplace:%i" % (estimated_num_seqs[data] // 1000)
+    align_dataset["seq_order_seq_lens_file"] = "/u/zeyer/setups/switchboard/dataset/data/seq-lens.train.txt.gz"
+
+  default_dataset = {"class": "my_default_dataset"}
+
+  d = {
+    "class": "MetaDataset",
+    "datasets": {"default": default_dataset, "align": align_dataset},
+    "data_map": {
+      "data": ("default", "data"),
+      "alignment": ("align", "data"),
+    },
+    "seq_order_control_dataset": "align",  # it must support get_all_tags
+  }
+  return d
+  pass
+
+
+class Stage:
+  def __init__(self, make_net,
+               stage_epochs: int,
+               fixed_path: bool,
+               alignment_topology: Topology,
+               name: str = None):
+    self.make_net = make_net
+    self.stage_epochs = stage_epochs
+    self.fixed_path = fixed_path  # False if full_sum and True if fixed_path
+    self.alignment_topology = alignment_topology  # rna, rnnt or ctc topology
+    if name is None:
+      name = alignment_topology.name + f"_{'fixed_path' if fixed_path else 'full_sum'}"
+    self.name = name  # name used to overwrite the model name for the checkpoints
+
+
+class MultiStager:
+  """ Wrapper around Pretrain which enables Multi-Stage training"""
+  def __init__(self, stage_list: List[Stage]):
+    self.stage = stage_list[0]  # saves the stage we are on
+    self.index = 0  # index of current stage
+    # accumulative sum of the epochs
+    # so that they represent epoch up to which the stage lasts
+    for i in range(len(stage_list) - 1):
+      stage_list[i + 1].stage_epochs += stage_list[i].stage_epochs + 1  # accumulative sum of the epochs
+    self.stage_list = stage_list
+
+  def update(self, epoch: int):
+    """ Update model for the next stage if necessary"""
+    # Update context(hmm)
+    task = get_global_config().value("task", "train")
+    target = TargetConfig.global_from_config()
+    model = get_global_config().value("model", "net-model/network")
+    self.ctx = Context(task=task, target=target, model=model, name=self.stage.name,
```

Comment:
This is very bad. You should never introduce attribs in non-init functions. (Basic Python rules. I think PyCharm would also warn you about this, or not?)

Also, you should not access the global config in other functions. In the optimal case, it would never be accessed at all, and all relevant context information is passed in `__init__`. If needed, it might be used for default arguments in `__init__`. See also other code.